File: Longest Substring Without Repeating Characters.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int maxLength = 0;
        int left = 0; // Left side of the sliding window
        
        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);
            if (map.containsKey(currentChar)) {
                // Move the left side of the window to the right of the first occurrence of currentChar
                left = Math.max(left, map.get(currentChar) + 1);
            }
            map.put(currentChar, right); // Update the last seen index of the currentChar
            maxLength = Math.max(maxLength, right - left + 1); // Calculate the max length so far
        }
        
        return maxLength;
    }
}

File: longest_substring_without_repeating_characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charIndexMap = {}
        maxLength = 0
        start = 0
        
        for i, char in enumerate(s):
            # If the character is in the map and its index is within the current window
            if char in charIndexMap and charIndexMap[char] >= start:
                start = charIndexMap[char] + 1  # Move the start of the window
            charIndexMap[char] = i
            maxLength = max(maxLength, i - start + 1)  # Update the max length
        
        return maxLength
